multiprocess: Add -ipcbind option to bitcoin-node

ryanofsky opened the pull request:
Add `-ipcbind` option to `bitcoin-node` to make it listen on a unix socket and accept connections from other processes. The default socket path is `<datadir>/node.sock`, but this can be customized.

This option lets potential wallet, gui, index, and mining processes connect to the node and control it. See examples in #19460, #19461, and #30437.

Motivation for this PR, in combination with #30510, is be able to release a bitcoin core node binary that can generate block templates for a separate Stratum v2 mining service, like the one being implemented in https://github.com/Sjors/bitcoin/pull/48, that connects over IPC.

Other things to know about this PR:

- While the `-ipcbind` option lets other processes to connect to the `bitcoin-node` process, the only thing they can actually do after connecting is call methods on the [`Init`](https://github.com/bitcoin/bitcoin/blob/master/src/ipc/capnp/init.capnp#L17-L20) interface which is currently very limited and doesn't do much. But PRs [#30510](https://github.com/bitcoin/bitcoin/pull/30510), [#29409](https://github.com/bitcoin/bitcoin/pull/29409), and [#10102](https://github.com/bitcoin/bitcoin/pull/10102) expand the `Init` interface to expose mining, wallet, and gui functionality respectively.

- This PR is not needed for [#10102](https://github.com/bitcoin/bitcoin/pull/10102), which runs GUI, node, and wallet code in different processes, because [#10102](https://github.com/bitcoin/bitcoin/pull/10102) does not use unix sockets or allow outside processes to connect to existing processes. [#10102](https://github.com/bitcoin/bitcoin/pull/10102) lets parent and child processes communicate over internal socketpairs, not externally accessible sockets.

---

This PR is part of the [process separation project](https://github.com/bitcoin/bitcoin/issues/28722).

ryanofsky committed:
multiprocess: Add -ipcbind option to bitcoin-node

Add `-ipcbind` option to `bitcoin-node` to listen on an IPC socket and accept
connections from other processes. In the future, there will be an `-ipcconnect`
option added to `bitcoin-wallet` and `bitcoin-node` to allow wallet and gui
processes to connect to the node and access it.

Example usage:

    src/bitcoin-node -regtest -debug -ipcbind=unix
    src/bitcoin-wallet -regtest -ipcconnect=unix info
    src/bitcoin-gui -regtest -ipcconnect=unix
    src/bitcoin-mine -regtest -ipcconnect=unix

ryanofsky committed:
multiprocess: Add IPC connectAddress and listenAddress methods

Allow listening on and connecting to unix sockets.

ryanofsky committed:
multiprocess: Add unit tests for connect, serve, and listen functions

Sjors commented:
One thing that may be useful to clarify, IIUC, although you can connect, nothing can be done via these connections yet. That is what #30510 introduces specifically for the Mining interface.

ryanofsky commented:
> One thing that may be useful to clarify, IIUC, although you can connect, nothing can be done via these connections yet. That is what #30510 introduces specifically for the Mining interface.

Thank you, added this to the description

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/30509#issue-2425377062

> The default socket path is currently `<datadir>/sockets/bitcoin-node.sock`, but this can be customized.

I changed the default socket path to `<datadir>/node.sock` after running into some "exceeded maximum socket path length" errors. The maximum unix socket path length is only 107 characters, so I thought a shorter suffix might avoid problems on systems with long data directory or home directory paths.

Another possible approach could be to use `$XDG_RUNTIME_DIR/bitcoin-node.sock` as the default socket path, since `$XDG_RUNTIME_DIR` should resolve to a short path like `/run/user/1000`. But this could lead to surprises on systems with multiple data directories since the socket path is no longer inside the data directory, and it could cause security problems since the socket would no longer be shielded by permissions of the data directory.

Updated 3d9a8cb4e194399fa9b2ce8af6bd2a8b8ef970c6 -> efdc120795e7c3ce914ee0677c3b69803ea5de1e ([`pr/ipc-bind.1`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bind.1) -> [`pr/ipc-bind.2`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bind.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-bind.1..pr/ipc-bind.2)) with this change and a test simplification.

TheCharlatan commented on src/init/bitcoin-gui.cpp:
```diff
@@ -35,6 +35,7 @@ class BitcoinGuiInit : public interfaces::Init
     std::unique_ptr<interfaces::Echo> makeEcho() override { return interfaces::MakeEcho(); }
     interfaces::Ipc* ipc() override { return m_ipc.get(); }
     node::NodeContext m_node;
+    bool canListenIpc() override { return true; }
```
Just a question: Intuitively I was under the impression that only the node should be listening. In which scenario should the gui be able to listen as well? Do we need a bi-directional connection here for some callbacks?

TheCharlatan commented on src/test/ipc_test.cpp:
```diff
@@ -65,3 +91,53 @@ void IpcTest()
+    std::unique_ptr<interfaces::Init> init{std::make_unique<TestInit>()};
+    std::unique_ptr<ipc::Protocol> protocol{ipc::capnp::MakeCapnpProtocol()};
+    std::unique_ptr<ipc::Process> process{ipc::MakeProcess()};
+    {
```
Since multiple binds are supported, I would suggest to add a test case for binding and connecting on different addresses in the same process.

TheCharlatan commented on src/test/ipc_test.cpp:
```diff
@@ -65,3 +91,53 @@ void IpcTest()
+    std::unique_ptr<interfaces::Echo> remote_echo{remote_init->makeEcho()};
+    BOOST_CHECK_EQUAL(remote_echo->echo("echo test"), "echo test");
+    remote_echo.reset();
+    remote_init.reset();
```
Is there a reason for manually resetting these here? I guess in the other test resetting is required, because otherwise the thread will never join.

sedited reviewed (commented):
lgtm , I hope we can finally get this project under way now.

ryanofsky reviewed (commented):
Thanks for the review!

Updated efdc120795e7c3ce914ee0677c3b69803ea5de1e -> cf9aaf4b7f27b38bf201aa39c56c9967f3b2a904 ([`pr/ipc-bind.2`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bind.2) -> [`pr/ipc-bind.3`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bind.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-bind.2..pr/ipc-bind.3)) with suggestions

ryanofsky commented:
The test failure https://cirrus-ci.com/task/5102088831631360 in the new test which tests binding and connecting on [multiple sockets](https://github.com/bitcoin/bitcoin/pull/30509#discussion_r1707288885) turns out to reveal a corner case bug in libmultiprocess. The bug does not happen locally for me, but it does happen if I run the CI scripts inside a container, I think maybe because the container is 32 bit.

Basically what happens is the test disconnects and reconnects immediately, which causes an `mp::Connection` instance to be freed and a new one to be allocated right away. In the CI container, the new mp::Connection object is allocated at the same address as the previous object and this causes the client connection try to reuse a remote thread associated with the previous connection which no longer exists. This happens because the client tracks threads to run on in a [map](https://github.com/chaincodelabs/libmultiprocess/blob/a9e16da55e880e5a0aed5008307ea56edc33fbd1/include/mp/proxy-io.h#L529) `std::map<Connection*, ProxyClient<Thread>> request_threads;` which is keyed by Connection pointer, with the assumption that that two different connections will not have the same pointer. Which is not the case when this test case is run in the CI container. I started implementing a fix for this which cleans up the map entry when the connection is destroyed, which is probably a good thing to do anyway so the map does not have stale entries in it.

ryanofsky committed:
depends: Update libmultiprocess library for CustomMessage function and ThreadContext bugfix

The CustomMessage functions allow simplifying custom IPC type code, and the
bugfix is needed to prevent in a crash in a new test which creates and destroys
connections in a loop. Upstream PRs are:

https://github.com/chaincodelabs/libmultiprocess/pull/105 types: Add Custom{Build,Read,Pass}Message hooks
https://github.com/chaincodelabs/libmultiprocess/pull/106 Bugfix: Clean up ThreadContext pointers when Connection is destroyed

ryanofsky commented:
Rebased cf9aaf4b7f27b38bf201aa39c56c9967f3b2a904 -> 4137419b0170fa1e9ff9daacde57933f7c995b76 ([`pr/ipc-bind.3`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bind.3) -> [`pr/ipc-bind.4`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bind.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-bind.3-rebase..pr/ipc-bind.4)) with fix for new test that was failing when run in the multiprocess CI container https://github.com/bitcoin/bitcoin/pull/30509#issuecomment-2276739868

Sjors commented:
`<datadir>/node.sock` seems fine, although on macOS with the longest permitted username you would go over the limit: `/Users/willem-alexanderclausgeorgeferdinand-konder-der-nederlanden-jonkheer-van-amsberg-67/Library/Application\ Support/Bitcoin/testnet4/bitcoin-node.sock`

ryanofsky commented:
> `<datadir>/node.sock` seems fine, although on macOS with the longest permitted username you would go over the limit: `/Users/willem-alexanderclausgeorgeferdinand-koning-der-nederlanden-jonkheer-van-amsberg-67/Library/Application\ Support/Bitcoin/testnet4/bitcoin-node.sock`

Ha! It is pretty easy to go over the socket path limit on linux too since it is only something like 100 characters. I think it might actually be possible to work around this by forking the process, changing the working directory, opening the socket with a relative path, and then passing back a file descriptor the parent process, but not sure about that and hopefully it would not be necessary.

Sjors commented on src/ipc/capnp/protocol.cpp:
```diff
@@ -51,6 +53,14 @@ class CapnpProtocol : public Protocol
+    void listen(int listen_fd, const char* exe_name, interfaces::Init& init) override
+    {
+        startLoop(exe_name);
+        if (::listen(listen_fd, 5 /* backlog */) != 0) {
```
52917ad4be1ca574543e854fdc154e88d65ce57f: nit `/*backlog=*/5`

Sjors commented on src/ipc/process.cpp:
```diff
@@ -54,7 +60,100 @@ class ProcessImpl : public Process
+    int bind(const fs::path& data_dir, const std::string& exe_name, std::string& address, std::string& error) override;
 };
+
+static bool ParseAddress(std::string& address,
```
52917ad4be1ca574543e854fdc154e88d65ce57f: this could use a test.

Or maybe we can introduce something akin to `CNetAddr` for socket addresses? Fortunately this input can only be provided by the user, so we don't have to be overzealous about sanitising.

Sjors reviewed (commented):
ACK 4137419b0170fa1e9ff9daacde57933f7c995b76

I only lightly looked at the new `ProcessImpl` method bodies.

I tested that I'm able to connect the bitcoin-mine demo #30437, but that (as expected) it immediately fails with a "remote exception: Method not implemented".

ryanofsky reviewed (commented):
Updated 4137419b0170fa1e9ff9daacde57933f7c995b76 -> af24810eeed397c2fe5f61ef9769e1b7ee0ecdf9 ([`pr/ipc-bind.4`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bind.4) -> [`pr/ipc-bind.5`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bind.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-bind.4..pr/ipc-bind.5)) with suggested changes simplifying error handling and adding more tests.

Sjors commented on src/init.cpp:
```diff
@@ -1200,6 +1204,16 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        for (std::string address : gArgs.GetArgs("-ipcbind")) {
+            std::string error;
+            if (!ipc->listenAddress(address, error)) {
+                return InitError(strprintf(Untranslated("Unable to bind to IPC address '%s'. %s"), address, error));
```
Works now.

Sjors commented:
re-ACK af24810eeed397c2fe5f61ef9769e1b7ee0ecdf9

TheCharlatan commented on src/common/args.cpp:
```diff
@@ -617,6 +617,9 @@ std::string ArgsManager::GetHelpMessage() const
             case OptionsCategory::RPC:
                 usage += HelpMessageGroup("RPC server options:");
                 break;
+            case OptionsCategory::IPC:
```
I think this should be added to the `rpc_help.py` tests.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30509#discussion_r1733464897

> I think this should be added to the `rpc_help.py` tests.

Is there mistake in this suggestion? This shouldn't be related to RPC

TheCharlatan replied:
Huh, no idea what I was thinking here, but `rpc_help.py` is obviously wrong. Maybe I copy-pasted a wrong value? I think where it could go is the `system.cpp` fuzz tests.

TheCharlatan commented on src/test/ipc_test.cpp:
```diff
@@ -65,3 +91,67 @@ void IpcTest()
+    std::unique_ptr<interfaces::Init> init{std::make_unique<TestInit>()};
+    std::unique_ptr<ipc::Protocol> protocol{ipc::capnp::MakeCapnpProtocol()};
+    std::unique_ptr<ipc::Process> process{ipc::MakeProcess()};
+
```
I think it would be good to sanity-check that invalid addresses make `bind` and `connect` throw:
```diff
diff --git a/src/test/ipc_test.cpp b/src/test/ipc_test.cpp
index f6addb2697..a5f6efbfb7 100644
--- a/src/test/ipc_test.cpp
+++ b/src/test/ipc_test.cpp
@@ -16,0 +17 @@
+#include <stdexcept>
@@ -121,0 +123,4 @@ void IpcSocketTest(const fs::path& datadir)
+    std::string invalid_bind{"invalid:"};
+    BOOST_CHECK_THROW(process->bind(datadir, "test_bitcoin", invalid_bind), std::invalid_argument);
+    BOOST_CHECK_THROW(process->connect(datadir, "test_bitcoin", invalid_bind), std::invalid_argument);
+
```

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30509#discussion_r1733490705

> I think it would be good to sanity-check that invalid addresses make `bind` and `connect` throw:

Thanks added suggested tests.

sedited reviewed (approved):
ACK af24810eeed397c2fe5f61ef9769e1b7ee0ecdf9

This looks good to me. Thanks for following up on my earlier suggestions, I left a couple more. It would be nice to add some functional tests too, but that is easily done in a follow-up.

ryanofsky commented on src/test/ipc_test.cpp:
```diff
@@ -65,3 +91,67 @@ void IpcTest()
+    //   Address 'unix' path '"/tmp/test_common_Bitcoin Core/ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff/test_bitcoin.sock"' exceeded maximum socket path length
+    std::vector<std::string> addresses{
+       strprintf("unix:%s", TempPath("bitcoin_sock0_XXXXXX")),
+       strprintf("unix:%s", TempPath("bitcoin_sock1_XXXXXX")),
```
re: https://github.com/bitcoin/bitcoin/pull/30509#discussion_r1732716844

> Nit: There is a problem with the indentation here.

Thanks! Fixed

ryanofsky reviewed (commented):
Rebased af24810eeed397c2fe5f61ef9769e1b7ee0ecdf9 -> e225f7cbc3e6842f8e7f1c482c2aacd810e99c1b ([`pr/ipc-bind.5`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bind.5) -> [`pr/ipc-bind.6`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bind.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-bind.5-rebase..pr/ipc-bind.6)) implementing suggestions and updating to use cmake

re: https://github.com/bitcoin/bitcoin/pull/30509#pullrequestreview-2263113816

>  It would be nice to add some functional tests too, but that is easily done in a follow-up.

Yeah it is a little harder to do now because this is only adding a bitcoin-node socket listening option, and there isn't an executable right now that connects to this socket. But #30437 would be an easier place to add this because it adds an executable that can connect.

sedited reviewed (approved):
Re-ACK e225f7cbc3e6842f8e7f1c482c2aacd810e99c1b

Sjors commented:
re-utACK e225f7cbc3e6842f8e7f1c482c2aacd810e99c1b

I plan to test this with https://github.com/Sjors/bitcoin/pull/48 but atm the rebases involved are a bit too much headache.

ryanofsky commented:
Rebased e225f7cbc3e6842f8e7f1c482c2aacd810e99c1b -> 30073e6b3a24cbe417c45cd5df6a3a2de0251e9d ([`pr/ipc-bind.6`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bind.6) -> [`pr/ipc-bind.7`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-bind.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-bind.6-rebase..pr/ipc-bind.7)) with no changes other than updating system.cpp fuzz test

sedited reviewed (approved):
Re-ACK 30073e6b3a24cbe417c45cd5df6a3a2de0251e9d

itornaza reviewed (approved):
Code review ACK 30073e6b3a24cbe417c45cd5df6a3a2de0251e9d

Followed step by step the implementation through the code and especially the `src/ipc/*` changed files and it looks solid. Great that it provides yet another step to facilitate the stratum v2 integration into core.

achow101 commented:
ACK 30073e6b3a24cbe417c45cd5df6a3a2de0251e9d

Sjors commented:
Post-merge re-utACK 30073e6b3a24cbe417c45cd5df6a3a2de0251e9d

ryanofsky commented on src/ipc/process.cpp:
```diff
@@ -54,7 +60,95 @@ class ProcessImpl : public Process
+    }
+    int connect_error = errno;
+    if (::close(fd) != 0) {
+        LogPrintf("Error closing file descriptor %i '%s': %s\n", fd, address, SysErrorString(errno));
```
> For new code

Thanks, this code was written 2018-08-23

l0rinc commented on src/ipc/protocol.h:
```diff
@@ -25,12 +25,38 @@ class Protocol
+    //! The optional `ready_fn` callback will be called after the event loop is
+    //! created but before it is started. This can be useful in tests to trigger
+    //! client connections from another thread as soon as the event loop is
+    //! available, but should not be neccessary in normal code which starts
```
nit: this introduced a lint-spelling warning:
> % test/lint/lint-spelling.py       
src/ipc/protocol.h:57: neccessary ==> necessary
